grapple/settings.py

```diff
@@ -0,0 +1,84 @@
+"""
+Settings for Wagtail Grapple are all namespaced in the GRAPPLE setting.
+For example your project's `settings.py` file might look like this:
+GRAPPLE = {
+    'APPS': ['home'],
+    'ADD_SEARCH_HIT': True,
+}
+This module provides the `grapple_settings` object, that is used to access
+Wagtail Grapple settings, checking for user settings first, then falling
+back to the defaults.
+"""
+from django.conf import settings
+from django.test.signals import setting_changed
+
+
+DEFAULTS = {
+    "APPS": ["home"],
+    "AUTO_CAMELCASE": True,
+    "ALLOWED_IMAGE_FILTERS": None,
+    "EXPOSE_GRAPHIQL": False,
+    "ADD_SEARCH_HIT": False,
+    "PAGE_SIZE": 10,
+    "MAX_PAGE_SIZE": 100,
+}
+
+
+class GrappleSettings:
+    """
+    A settings object that allows Wagtail Grapple settings to be accessed as
+    properties. For example:
+        from grapple.settings import grapple_settings
+        print(grapple_settings.APPS)
+    Note:
+    This is an internal class that is only compatible with settings namespaced
+    under the GRAPPLE name. It is not intended to be used by 3rd-party
+    apps, and test helpers like `override_settings` may not work as expected.
+    """
+
+    def __init__(self, user_settings=None, defaults=None):
+        if user_settings:
+            self._user_settings = user_settings
+        self.defaults = defaults or DEFAULTS
+        self._cached_attrs = set()
+
+    @property
+    def user_settings(self):
+        if not hasattr(self, "_user_settings"):
+            self._user_settings = getattr(settings, "GRAPPLE", {})
+        return self._user_settings
+
+    def __getattr__(self, attr):
+        if attr not in self.defaults:
+            raise AttributeError("Invalid Grapple setting: '%s'" % attr)
+
+        try:
+            # Check if present in user settings
+            val = self.user_settings[attr]
+        except KeyError:
```

Comment:
This is an excellent solution for the single setting approach!

We have a "beta"  pypi classifier (and by semver standards not being a full major version) which in a way gives us freedom to introduce breaking changes, I feel that the package is used by enough production sites that we should follow a 2 release deprecation process.

As such, would be good to try to support the old settings and we will make a note in the release notes about them being deprecated and pending removal in the following release.

so I would add a bit more code here that tries to get the value from the separate settings

Reply:
ok, I'll add that logic.

Reply:
@zerolab I just added `deprecated` and `removed` settings support. Deprecated settings will log a warning and removed settings will raise a `RuntimeError`.